alf/algorithms/sac_algorithm.py

```diff
@@ -228,102 +261,252 @@ def __init__(self,
-                if self._dqda_clipping:
-                    dqda = torch.clamp(dqda, -self._dqda_clipping,
-                                       self._dqda_clipping)
-                loss = 0.5 * losses.element_wise_squared_loss(
-                    (dqda + action).detach(), action)
-                loss = loss.sum(list(range(1, loss.ndim)))
-                return loss
-
-            actor_loss = nest.map_structure(actor_loss_fn, dqda, action)
-            alpha = torch.exp(self._log_alpha).detach()
-            actor_loss += alpha * log_pi
+        return self.predict_step(time_step, state, epsilon_greedy=1.0)
+
+    def _compute_critics(self, critic_net, observation, action, critics_state):
+        if self._act_type == ActionType.Continuous:
+            observation = (observation, action)
+        elif self._act_type == ActionType.Mixed:
+            observation = (observation, action[1])  # continuous action
+        # discrete/mixed: critics shape [B, replicas, num_actions]
+        # continuous: critics shape [B, replicas]
+        critics, critics_state = critic_net(observation, state=critics_state)
+        return critics, critics_state
+
+    def _actor_train_step(self, exp: Experience, state, action, critics,
+                          log_pi):
+        if self._act_type == ActionType.Discrete:
+            # Pure discrete case doesn't need to learn an actor network
+            return (), ()
+
+        if self._act_type == ActionType.Continuous:
+            critics, critics_state = self._compute_critics(
+                self._critic_networks, exp.observation, action, state)
         else:
-            critics, critics_state = self._critic_networks(
-                exp.observation, state=state.critics)
-
-            base_action_dist = dist_utils.get_base_dist(action_distribution)
-            assert isinstance(base_action_dist, td.categorical.Categorical),  \
-                 ("Only `Categorical` " + "was supported, received:" + str(
-                        type(base_action_dist)))
-
-            log_action_probs = base_action_dist.logits.squeeze(1)
-
-            target_q_value = critics.min(dim=1)[0].detach()
-            alpha = torch.exp(self._log_alpha)
-            actor_loss = torch.exp(log_action_probs) * (
-                alpha.detach() * log_action_probs - target_q_value)
-            actor_loss = actor_loss.mean(list(range(1, actor_loss.ndim)))
-
-        state = SacActorState(critics=critics_state)
-        info = SacActorInfo(loss=LossInfo(loss=actor_loss, extra=actor_loss))
-
-        return state, info
+            # use the provided critics for discrete/mixed
+            critics_state = ()
+
+        if self._act_type == ActionType.Mixed:
+            # pick values according to discrete actions
+            critics = self._select_q_value(action[0], critics).squeeze(-1)
+            action = action[1]
+
+        target_q_value = critics.min(dim=1)[0]
+
+        dqda = nest.pack_sequence_as(
```

Comment:
This is worth to be a helper function:  ``grad(nest, objective)``. It's used at several places: ddpg, sarsa, sac.